src/devices/Board/Board.cs

```diff
@@ -0,0 +1,407 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Device.Gpio;
+using System.Device.I2c;
+using System.Device.Pwm;
+using System.Device.Spi;
+using System.IO;
+using System.Linq;
+using System.Text;
+
+namespace Iot.Device.Board
+{
+    public abstract class Board : MarshalByRefObject, IDisposable
+    {
+        private readonly PinNumberingScheme _defaultNumberingScheme;
+        private readonly object _pinReservationsLock;
+        private readonly Dictionary<int, List<PinReservation>> _pinReservations;
+        private bool _initialized;
+        private bool _disposed;
+
+        protected Board(PinNumberingScheme defaultNumberingScheme)
+        {
+            _defaultNumberingScheme = defaultNumberingScheme;
+            _pinReservations = new Dictionary<int, List<PinReservation>>();
+            _pinReservationsLock = new object();
+            _initialized = false;
+            _disposed = false;
+        }
+
+        ~Board()
+        {
+            Dispose(false);
+        }
+
+        public event Action<string, Exception> LogMessages;
+
+        protected bool Initialized
+        {
+            get
+            {
+                return _initialized;
+            }
+        }
+
+        protected bool Disposed
+        {
+            get
+            {
+                return _disposed;
+            }
+        }
+
+        public PinNumberingScheme DefaultPinNumberingScheme
+        {
+            get
+            {
+                return _defaultNumberingScheme;
+            }
+        }
+
+        protected void Log(string message, Exception exception = null)
+        {
+            LogMessages?.Invoke(message, null);
+        }
+
+        /// <summary>
+        /// Converts pin numbers from one numbering scheme to another.
+        /// </summary>
+        /// <param name="pinNumber">Pin number to convert</param>
+        /// <param name="inputScheme">The numbering scheme of the input pin</param>
+        /// <param name="outputScheme">The desired numbering scheme</param>
+        /// <returns>The converted pin number</returns>
+        public abstract int ConvertPinNumber(int pinNumber, PinNumberingScheme inputScheme, PinNumberingScheme outputScheme);
+
+        /// <summary>
+        /// Reserves a pin for a specific usage. This is done automatically if a known interface (i.e. GpioController) is
+        /// used to open the pin, but may be used to block a pin explicitly, i.e. for UART.
+        /// </summary>
+        /// <param name="pinNumber">The pin number, in the boards default numbering scheme</param>
+        /// <param name="usage">Intended usage of the pin</param>
+        /// <param name="owner">Class that owns the pin (use "this")</param>
+        public virtual void ReservePin(int pinNumber, PinUsage usage, object owner)
```

Comment:
I'd prefer this was not part of the initial version of abstraction, this needs more thought. Perhaps this is needed long term and this is how the API would look like but I'd prefer we did it separately as this PR is already large enough for this rather niche concept

Reply:
I have to check whether it can be removed easily. I think the method is internally used already.

Reply:
This is one of the main features of the new approach, actually. It manages the pin usages and keeps track of who uses which pin for what purpose. This is public to also allow reserving pins directly from a client (i.e. for UART, which is not currently managed here)